kubectl-cloudflow/cmd/deploy.go

```diff
@@ -81,22 +105,156 @@ the stored credentials.
 
 You can update the credentials with the "update-docker-credentials" command.
 `,
-		Example: `kubectl cloudflow deploy registry.test-cluster.io/cloudflow/sensor-data-scala:292-c183d80 valid-logger.log-level=info valid-logger.msg-prefix=valid`,
-		Args:    validateDeployCmdArgs,
-		Run:     deployOpts.deployImpl,
+		Example: `kubectl cloudflow deploy -i registry.test-cluster.io/cloudflow/sensor-data-scala:292-c183d80 valid-logger.log-level=info valid-logger.msg-prefix=valid`,
+		PreRun: func(cmd *cobra.Command, args []string) {
+			if cmd.Flag("image").Changed == false && cmd.Flag("blueprint").Changed == false {
+				util.LogAndExit("Please specify either the blueprint file or the image path")
+			}
+			if (cmd.Flag("image").Changed == true) && (cmd.Flag("blueprint").Changed == true) {
+				util.LogAndExit("%s", "Cannot specify both blueprint and image in command")
+			}
+			if cmd.Flag("image").Changed == true {
+				fmt.Printf("Deploying image %s\n", cmd.Flag("image").Value.String())
+			} else if cmd.Flag("blueprint").Changed == true {
+				fmt.Printf("Deploying blueprint %s\n", cmd.Flag("blueprint").Value.String())
+			}
+		},
+		Run: deployOpts.deployImpl,
 	}
+	deployOpts.cmd.Flags().StringVarP(&deployOpts.blueprintFilePath, "blueprint", "b", "", "blueprint file path.")
+	deployOpts.cmd.Flags().StringVarP(&deployOpts.imagePath, "image", "i", "", "docker image path.")
 	deployOpts.cmd.Flags().StringVarP(&deployOpts.username, "username", "u", "", "docker registry username.")
 	deployOpts.cmd.Flags().StringVarP(&deployOpts.password, "password", "p", "", "docker registry password.")
 	deployOpts.cmd.Flags().BoolVarP(&deployOpts.passwordStdin, "password-stdin", "", false, "Take the password from stdin")
 	deployOpts.cmd.Flags().StringArrayVar(&deployOpts.volumeMounts, "volume-mount", []string{}, "Accepts a key/value pair separated by an equal sign. The key should be the name of the volume mount, specified as '[streamlet-name].[volume-mount-name]'. The value should be the name of an existing persistent volume claim.")
+	deployOpts.cmd.Flags().StringToIntVar(&deployOpts.replicasByStreamletName, "scale", map[string]int{}, "Accepts key/value pairs for replicas per streamlet")
 
 	rootCmd.AddCommand(deployOpts.cmd)
 }
 
 func (opts *deployOptions) deployImpl(cmd *cobra.Command, args []string) {
 	version.FailOnProtocolVersionMismatch()
 
-	imageRef := args[0]
+	if cmd.Flag("image").Changed == true {
+		deployImage(cmd, opts, args)
+	} else {
+		deployBlueprint(cmd, opts, args)
+	}
+}
+
+func deployBlueprint(cmd *cobra.Command, opts *deployOptions, args []string) {
+	// read blueprint file and load its content
+	blueprintFile := cmd.Flag("blueprint").Value.String()
+	contents, err := fileutils.GetFileContents(blueprintFile)
+	contents = strings.TrimSpace(contents)
+	if err != nil {
+		util.LogAndExit("Failed to fetch blueprint contents from %s", blueprintFile)
+	}
+
+	blueprint, err := verify.VerifyBlueprint(contents)
```

Comment:
+1

Reply:
At this point you must have already pulled all images, read all labels. It seems that CreateApplicationSpecFromBlueprintAndImages pulls images again and gets labels again? This can probably be improved.

Reply:
@RayRoestenburg we don't pull images in `CreateApplicationSpecFromBlueprintAndImages` - I use the information from the already filled up `blueprint` which is passed to the function.

Reply:
It uses getStreamletDescriptorsAndImageInformation, which does the image list again, which is probably already done before?